### ISSUE TITLE
RMTPP on continous time-series data

### ISSUE THREAD

**Issue by abinashsinha330** [musically-ut/tf_rmtpp#6]
@musically-ut I see that each sequence in _event.txt file and _time.txt file which are the input formats start from the beginning time.

What if we have a continuous time-series data which has around a billion traces with me starting from beginning till some time and each trace is annotated with a certain label?

How are you planning to sample that kind of data for pre-processing to get the desired input .txt files in your format?

**Comment by musically-ut** [comment]
In that case, one can pick an arbitrary window of time `T` from any starting point for training and adjust all time stamps within the window such that the starting point is `t = 0` while preserving all inter-event times.

Does that make sense?

**Comment by abinashsinha330** [comment]
so if I have a very long time-series starting at time t = 0s till t = 9000s and say there are a total of a billion events that occurred. Each event can be of either of three types. Can I take a window of let's say 10s on a chronological basis consecutively and construct the sets of sequences of events and corresponding sequences of times of occurrence of event.

For example,

Time of occurrence (in s)    Event Type
           0                                   High
           2                                   Low
           2.3                                Low
           2.5                                   High
           3.4                                   Low
           3.45                                Low
           4                                   High
           4.10                                   Low
           5                                Low
           6                                    High
          .                                      .
          .                                      .
.. and so on


So, is it okay if I create sequences with time window of let's say 3 seconds like below:

**time.txt**:
Sequence 1: 0 2 2.3 2.5
Sequence 2: 3.4  3.45  4  4.10  5  6
Sequence 3:
.
.
and so on

I will adjust the time sequences here as part of pre-processing.

**event.txt**:
Sequence 1: High Low Low High
Sequence 2: Low Low High Low Low High
Sequence 3:
.
.

Is this something you are suggesting to do? Also have you tried with such a kind of data? 
Because I see sequences of twitter related data where each sequence is for a particular user and something like for taxi data.

**Comment by musically-ut** [comment]
Our datasets usually have a key (e.g. `taxi_id`, `twitter_user_id`, or `stock`) by which we could separate out the time series into independent chunks of varying length, but all starting at time `t=0`.

If I understand you correctly, you are trying to predict the future values for one given time series only. In that case, the scheme you describe is the way to go. I do not remember whether the code adjusts the first event to be at zero or not, but you can make that adjustment yourself while generating `time.txt`, i.e., instead of:

**time.txt:**
```
Sequence 1: 0 2 2.3 2.5
Sequence 2: 3.4 3.45 4 4.10 5 6
Sequence 3:
.
.
```

generate:

**time.txt:**
```
Sequence 1: 0 2 2.3 2.5
Sequence 2: 0 0.05 0.6 0.7 1.4 2.6
Sequence 3:
.
.
```

Does that make sense?

**Comment by abinashsinha330** [comment]
Offsetting the times of events wouldn't it impact the underlying semantics of event generation process. I may be wrong though but what I see is each sequence in your case starts from start which is not so in my case.

**Comment by musically-ut** [comment]
The model itself is time-invariant because it only handles the time difference between events, and never the absolute time itself. The normalization to `t=0` prevents very large numbers from appearing and causing over/underflow errors accidentally, e.g.

```python
t1 = np.array([10**8], dtype=np.float32)
t2= np.array([10**8 + 1], dtype=np.float32)
t2 - t1
##>>> array([0.], dtype=float32)
```

**Comment by abinashsinha330** [comment]
I know it handles the time difference but I want to understand at a very fundamental level. Let's say my time-series which starts at t = 0 and goes until 9000s and even generation frequency is normally pretty high you will see events occurring at every 0.5s or something but there are cases where there were no events let's say in between 4000s and 4050s (50s window in which no even is occurring).

How is your RNN-based temporal process able to understand that during this window (between 4000s and 4050s) it had no event?

**Comment by abinashsinha330** [comment]
I think this is not what you should have recommended for offsetting the timings of events. You have used the notion of having ***ENTITY*** while creating your sequences. What if there is no notion of entity in the sequences?

**Comment by musically-ut** [comment]
> How is your RNN-based temporal process able to understand that during this window (between 4000s and 4050s) it had no event?

We calculate the likelihood of a sequence of events (i.e., their interarrival times) using the theory of Temporal Point Processes. Here are some lecture notes for a quick introduction to Point processes: http://courses.mpi-sws.org/hcml-ws18/lectures/TPP.pdf

> What if there is no notion of entity in the sequences?

In this case, to leverage training on GPUs and to alleviate the problem of vanishing/exploding gradients [1], one usually breaks up the single long sequence to create a _batch_ of sequences to train the network on.

Perhaps that you have in mind is the ideal setting where one would want to train with just one very long sequence, potentially repeated with random initial hidden states [2] to create a large batch.

This may generate a network which overfits to this particular sequence (by potentially memorizes parts of it). A model which would generalize better would be one which intuitively understands the _law_ governing the time intervals. We assume that this law, or the underlying dynamic process, which governs the next event is time-invariant and its dependence on the initial state decays _rapidly_. For example, if you are assuming that the underlying generative process is Poisson-like, then we know that the dependence on the initial state decays exponentially [3]. Similar decay will be seen with Markovian assumptions.

To be able to generate a network which generalizes better, one way is to start the sequence at different points in time so that the network is not aware of the full history, but still should be able to predict the future time events after having observed some history (in case of Poisson, just a logarithmically short period of history, but longer as the modelled process becomes more complex). Further, we may not want to show the network the entire sequence right to the end point after starting it at a random initial point because of the fear that it may again memorize the last parts. 

Therefore, **one way** to train a network which generalizes better is to train it on random windows of time; thus faking the notion of ENTITY.

Note that though this method has the added advantage of being able to exploit GPU parallelization and faster training iterations, there are other ways of making models generalize better: e.g. by restricting the model size, by regularization [4], etc.
 
Hope that helps.

[1]: The by-the-books method of training RMTPP does not suffer from vanishing/exploding gradients because it uses truncated BPTT.
[2]: RMTPP initializes the hidden states as zero for simplicity.
[3]: https://arxiv.org/abs/1905.12781
[4]: See Proposition 2: https://arxiv.org/pdf/1805.09360.pdf

**Comment by abinashsinha330** [comment]
No I do not want to employ the whole very long sequence since it is very long and doesn't make sense in terms of making the model implementable in production and yes I am trying to build a pseudo notion of entity in order to disintegrate the whole time-series into multiple sequences where each first event in sequence starts at around same time...

What my question basically is it going to matter if those sequences of events be chronologically increasing? Let's say

First event of 1st sequence occurs at t=0 and its last event occurs at t = 20s
First event of 2nd sequence occurs at t=21s and its last event occurs at t = 24s
First event of 3rd sequence occurs at t=25s and its last event occurs at t = 39s
First event of 4th sequence occurs at t=40s and its last event occurs at t = 45s
...
...
...
First event of nth sequence occurs at t=1000s and its last event occurs at t = 1007s

I understand the model and marked temporal point process but would feeding sequences in this manner would not be able to capture the semantics of when does each event happen on absolute scale right?

The case of your financial data you just fed the whole long sequence which was the only sequence which I plan not to do.

I might be asking a very basic question here. I just want to understand it at the basic level.

**Comment by musically-ut** [comment]
> What my question basically is it going to matter if those sequences of events be chronologically increasing? 

No. It does not even matter if the sequences overlap as long as they are "representative" of the underlying data which the network is supposed to model. You can dynamically generate new batches with random start and end time points for each epoch if you want, or you can keep data split in neat disjoint windows as you have done here.

> [...]  would **feeding** sequences in this manner would not be able to capture the semantics of when does each event happen on absolute scale right?

It is unclear what 'feeding' here means. For each training iteration, we _feed_ in `BATCH_SIZE` number of sequences. 

This batch in case of the paper, is generated by picking sequences at random. No dependence is assumed between two sequences in the same batch or across batches, and the hidden state is reset to zero at the beginning of every sequence. Hence, there is no notion of capturing any sign of a previous sequence, let alone abosolute time inherent in the model.

Does that help clarify something?

**Comment by abinashsinha330** [comment]
I would be grateful if you could clarify one thing... Why did you just use a single long sequence of 624,149 events for Financial Transaction Dataset rather than creating random sequences out of this time-series of 624,149 chronological events?

**Comment by musically-ut** [comment]
Sorry, it is unclear which dataset you are referring to?

**Comment by abinashsinha330** [comment]
It is in Section 6.3 of paper where real data was tested..

**Comment by musically-ut** [comment]
From what I recall, and what I see in the repository, we did [break up the event sequence](https://github.com/musically-ut/tf_rmtpp/tree/master/data/real/book_order). I believe the effect of the edge predictions (artefacts of breaking the sequence in non-overlapping parts) was negligible.

**Comment by abinashsinha330** [comment]
I think training would be very slow and problematic if the number of classes is very huge... It can be extended to include sample softmax for event predictions in that case isn't it?

**Comment by musically-ut** [comment]
> sample softmax for event predictions in that case isn't it?

Probably. I tend to prefer Hierarchical Softmax or explicit extreme classification techniques, but the final choice differs from scenario to scenario.

> [...] on what basis did you decide to 1e-6 to be the least the log_Pr_event can go while calculating the cross-entropy for the event predictions?

The motivation behind this was primarily numerical stability.

----

Shall we close this issue if your primary question has been answered?

 We can continue our discussions on new issues if you have any more questions.

**Comment by abinashsinha330** [comment]
Thanks for responding and helping out.
